Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs

```diff
@@ -416,6 +417,47 @@ public AndConstraint<TAssertions> NotBeAssignableTo(Type type, string because =
         return new AndConstraint<TAssertions>((TAssertions)this);
     }
 
+    /// <summary>
+    /// Allows combining one or more assertions using the other assertion methods that this library offers on an instance of <typeparamref name="T"/>.
+    /// </summary>
+    /// <remarks>
+    /// If multiple assertions executed by the <paramref name="assertion"/> fail, they will be raised as a single failure.
+    /// </remarks>
+    /// <param name="assertion">The element inspector which must be satisfied by the <typeparamref name="TSubject" />.</param>
+    /// <returns>An <see cref="AndConstraint{T}" /> which can be used to chain assertions.</returns>
+    /// <exception cref="ArgumentNullException"><paramref name="assertion"/> is <see langword="null"/>.</exception>
+    public AndConstraint<TAssertions> Satisfy<T>(Action<T> assertion)
+        where T : TSubject
+    {
+        Guard.ThrowIfArgumentIsNull(assertion, nameof(assertion), "Cannot verify an object against a <null> inspector.");
+
+        Execute.Assertion
+            .ForCondition(Subject is T)
+            .WithDefaultIdentifier(Identifier)
+            .FailWith("Expected {context:object} to be assignable to {0}{reason}, but {1} is not.", typeof(T), Subject.GetType());
+
+        string[] failuresFromInspector;
+
+        using (var assertionScope = new AssertionScope())
+        {
+            assertion((T)Subject);
```

Comment:
If `Satisfy` is used inside an `AssertionScope`, then the execution won't stop if `Subject is T` above fails, meaning we would send in a `null` to `action`.

```cs
object subject = null;

Action act = () =>
{
    using var _ = new AssertionScope();
    subject.Should().Satisfy<object>(x => x.Should().NotBeNull());
};
```

What we usually do to guard against this.
```cs
bool success = Execute.Assertion
    .ForCondition(...)
    .FailWith(...);

if (success)
{
    Execute.Assertion
        .ForCondition(...)
        .FailWith(...);
}
```

Reply:
I don't think I understand exactly what you mean.
Is it about using an assertion scope within `Satisfy` like this:
```cs
personAndAddressDto.Should().Satisfy<PersonAndAddressDto>(dto => {
    dto.Person.Name.Should().NotBeNullOrEmpty();
    using(new AssertionScope())
    {
        dto.Address.Should().Satisfy<AddressDto>(address => address.City.Should().NotBeNullOrEmpty());
    }
}
```

After I added the null-check, it seems to work as I would expect.

I'm not sure why anyone would add a nested assertion scope though, but of course it needs to be handled correctly.

Reply:
What @jnyrup means is that if you're new `Satisfy` is wrapped in an `AssertionScope` (just like you do yourself), the `FailWith` call with _not_ throw an exception and return `false` instead.

Reply:
I'm so sorry for the late reply. I will try to support that.

Reply:
Does this test verify the expected behavior (it passes with the current implementation)?

```cs
[Fact]
public void Using_assertion_scope_with_null_subject()
{
    // Arrange
    object subject = null;

    // Act
    Action act = () =>
    {
        using (new AssertionScope())
        {
            subject.Should().Satisfy<object>(x => x.Should().NotBeNull());
        }
    };

    // Assert
    act.Should().Throw<XunitException>()
        .WithMessage("Expected subject to be assignable to System.Object, but found <null>.");
}
```

If this is not what you mean, I need some help understanding the scenario you have in mind.

Reply:
Your `Using_assertion_scope_with_null_subject` reveals that using `Subject.GetType()` throws an NRE and needs to be `Subject?.GetType()`

If we also change `Typed_object_satisfied_against_incorrect_type_throws` to have 

```cs
Action act = () =>
{
    using var _ = new AssertionScope();
    personDto.Should().Satisfy<AddressDto>(dto => dto.Should().NotBeNull());
};
```

that reveals that `assertion((T)Subject);` throws `System.InvalidCastException: 'Unable to cast object of type 'PersonDto' to type 'AddressDto'.'`

We usually handle this by return the return value of the assertion chain and check if it was successful.

```cs
bool success = Execute.Assertion
    .ForCondition(Subject is not null)
    .WithDefaultIdentifier(Identifier)
    .FailWith("Expected {context:object} to be assignable to {0}{reason}, but found <null>.", typeof(T))
    .Then
    .ForCondition(Subject is T)
    .WithDefaultIdentifier(Identifier)
    .FailWith("Expected {context:object} to be assignable to {0}{reason}, but {1} is not.", typeof(T), Subject?.GetType());

if (success)
{
    string[] failuresFromInspector;

    using (var assertionScope = new AssertionScope())
    {
        assertion((T)Subject);
        failuresFromInspector = assertionScope.Discard();
    }

    if (failuresFromInspector.Length > 0)
    {
        string failureMessage = Environment.NewLine
            + string.Join(Environment.NewLine, failuresFromInspector.Select(x => x.IndentLines()));

        Execute.Assertion
            .WithDefaultIdentifier(Identifier)
            .WithExpectation("Expected {context:object} to match inspector, but the inspector was not satisfied:", Subject)
            .FailWithPreFormatted(failureMessage);
    }
}

return new AndConstraint<TAssertions>((TAssertions)this);
```

Reply:
I am unable to reproduce the exception with the suggested test.
Have you tested whether it throws an NRE?

I understand that the combination of `Execute.Assertion` and `AssertionScope` should behave in some specific way, but I don't know how that is.

In any case, I added the precondition check and the test, as suggested, even though there is no difference, regardless of what I do.

Could you please test it? I feel I'm wasting too much of your time because I don't understand the underlying problems you are trying to explain.

Reply:
Run `Using_assertion_scope_with_null_subject` with a debugger attached and for a moment remove
https://github.com/fluentassertions/fluentassertions/blob/8656682fe60df5ebff2cd6cc7a0d3c140e96786b/Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs#L15

When using an `AssertionScope` failing calls to `ForCondition` doesn't immediately cause the entire assertion to stop.
This is usually used for doing multiple assertions on different subjects and get all failures in the output.
E.g. if both of these assertions fail we get two failure messages instead of only the one from the first assertion.
```cs
using (new AssertionScope())
{
    a.Should().BeFoo();
    b.Should().BeBar();
}
```

As `AssertionScope`s continues the execution, we also internally in FluentAssertions need to guard against the execution continuing.

So when wrapping the call to `Satisfy` in an `AssertionScope` the assertion continues executing after `Subject is not null` fails, which leads to the NRE when calling `Subject!.GetType()` on a null `Subject`.
![image](https://github.com/fluentassertions/fluentassertions/assets/919634/e66899e2-415c-409e-9e9c-cf380e43be21)

The fix here is to change `Subject!.GetType()` to `Subject?.GetType()`

Your test `Using_assertion_scope_with_subject_satisfied_against_incorrect_type_throws` correctly exercises the case where  we continue after `Subject is T` fails and `if (success)` guards against casting `Subject` to `T`.
![image](https://github.com/fluentassertions/fluentassertions/assets/919634/bf43d4c0-33d1-4af1-ba0e-66b49685cd43)

Reply:
Ah, okay, now I'm starting to understand the problem.
The tests are not throwing the NREs, but internally, Fluent Assertions are not executing the assertion because it fails to handle the second precondition.

I mistakenly assumed `.Then` only got executed if the previous assertion succeeded.

Would it be better to split the checks instead of using `.Then`?

Reply:
Your fix with replacing `!` with `?` is exactly what I would do.